drivers/sensor/lsm6dsl/lsm6dsl_shub.c

```diff
@@ -0,0 +1,395 @@
+						     slv_cfg, 2) < 0) {
+			SYS_LOG_DBG("error writing embedded reg");
+			return -EIO;
+		}
+
+		slv_cfg[0] = value[cnt];
+		if (lsm6dsl_shub_write_embedded_regs(data,
+					LSM6DSL_EMBEDDED_SLV0_DATAWRITE,
+					slv_cfg, 1) < 0) {
+			SYS_LOG_DBG("error writing embedded reg");
+			return -EIO;
+		}
+
+		/* turn SH on */
+		lsm6dsl_shub_enable(data);
+		k_sleep(100); /* (TBD) based on odr */
+		lsm6dsl_shub_disable(data);
+
+		cnt++;
+	}
+
+	/* Put master in IDLE mode */
+	slv_cfg[0] = LSM6DSL_EMBEDDED_SLV0_WRITE_IDLE;
+	slv_cfg[1] = lsm6dsl_shub_sens_list[0].wai_addr;
+	slv_cfg[2] = LSM6DSL_EMBEDDED_SLVX_THREE_SENS;
+	if (lsm6dsl_shub_write_embedded_regs(data,
+					     LSM6DSL_EMBEDDED_SLV0_ADDR,
+					     slv_cfg, 3) < 0) {
+		SYS_LOG_DBG("error writing embedded reg");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+/*
+ * SLAVEs configurations:
+ *
+ *  - SLAVE 0: used for configuring the slave device
+ *  - SLAVE 1: used as data read channel to slave device
+ *  - SLAVE 2: used for generic reads while data channel is enabled
+ */
+static int lsm6dsl_shub_set_data_channel(struct lsm6dsl_data *data)
+{
+	u8_t slv_cfg[3];
+	u8_t slv_i2c_addr = lsm6dsl_shub_sens_list[0].i2c_addr[ext_i2c_addr];
+
+	/* SLV0 is used for generic write */
+	slv_cfg[0] = LSM6DSL_EMBEDDED_SLV0_WRITE_IDLE;
+	slv_cfg[1] = lsm6dsl_shub_sens_list[0].wai_addr;
+	slv_cfg[2] = LSM6DSL_EMBEDDED_SLVX_THREE_SENS;
+	if (lsm6dsl_shub_write_embedded_regs(data,
+					     LSM6DSL_EMBEDDED_SLV0_ADDR,
+					     slv_cfg, 3) < 0) {
+		SYS_LOG_DBG("error writing embedded reg");
+		return -EIO;
+	}
+
+	/* Set data channel for slave device */
+	slv_cfg[0] = (slv_i2c_addr << 1) | LSM6DSL_EMBEDDED_SLVX_READ;
+	slv_cfg[1] = lsm6dsl_shub_sens_list[0].out_data_addr;
+	slv_cfg[2] = lsm6dsl_shub_sens_list[0].out_data_len;
+	if (lsm6dsl_shub_write_embedded_regs(data,
+					     LSM6DSL_EMBEDDED_SLV1_ADDR,
+					     slv_cfg, 3) < 0) {
+		SYS_LOG_DBG("error writing embedded reg");
+		return -EIO;
+	}
+
+	/* turn SH on */
+	lsm6dsl_shub_enable(data);
+	k_sleep(100); /* (TBD) based on odr */
+
+	return 0;
+}
+
+int lsm6dsl_shub_read_external_chip(struct device *dev, u8_t *buf, u8_t len)
+{
+	struct lsm6dsl_data *data = dev->driver_data;
+
+	data->hw_tf->read_data(data, LSM6DSL_REG_SENSORHUB1, buf, len);
+
+	return 0;
+}
+
+int lsm6dsl_shub_init_external_chip(struct device *dev)
+{
+	struct lsm6dsl_data *data = dev->driver_data;
+	u8_t i;
+	u8_t chip_id = 0;
+	u8_t slv_i2c_addr;
+	u8_t slv_wai_addr = lsm6dsl_shub_sens_list[0].wai_addr;
+
+	for (i = 0; i < ARRAY_SIZE(lsm6dsl_shub_sens_list[0].i2c_addr); i++) {
+		slv_i2c_addr = lsm6dsl_shub_sens_list[0].i2c_addr[i];
+
+		if (slv_i2c_addr == 0) {
+			continue;
+		}
+
+		if (lsm6dsl_shub_read_slave_reg(data, slv_i2c_addr,
+						slv_wai_addr,
+						&chip_id, 1) < 0) {
+			SYS_LOG_DBG("failed reading external chip id");
+			return -EIO;
+		}
+		if (chip_id == lsm6dsl_shub_sens_list[0].wai_val) {
+			break;
+		}
+	}
+
+	if (i >= 2) {
```

Comment:
Is this check correct? Should you be comparing `chip_id` with something instead?